Adding manual contact tracing

shaunkhoo opened the pull request:
Manual contact tracing (using the same channel as digital contact tracing but without the app restrictions) implemented through the manual_tracing_traceable_fraction. If you are an app user, you essentially get two chances to be traced (which is what happens in reality). 

It is possible to differentiate manual contact tracing success rate by network (work vs random for example), but that will require more time.

roberthinch commented on tests/data/baseline_parameters.csv:
```diff
@@ -1,2 +1,2 @@
-rng_seed,param_id,n_total,mean_work_interactions_child,mean_work_interactions_adult,mean_work_interactions_elderly,daily_fraction_work,work_network_rewire,mean_random_interactions_child,sd_random_interactions_child,mean_random_interactions_adult,sd_random_interactions_adult,mean_random_interactions_elderly,sd_random_interactions_elderly,random_interaction_distribution,child_network_adults,elderly ...
-1,1,1000000,10,7,3,0.5,0.1,2,2,4,4,3,3,1,0.2,0.2,10,200,5,6,2.5,5.75,6,2.5,5.14,2.5,12,5,6,2,0.18,0.18,0.18,0.18,0.18,0.18,0.18,0.18,0.18,0.29,0.79,0.79,0.73,0.68,0.65,0.59,0.53,0.41,0.27,0.48,15,5,7452,9936,4416,4140,1104,552,8054000,7528000,8712000,8835000,8500000,8968000,7069000,5488000,3281000,0.002,0.71,0.74,0.79,0.87,0.98,1.11,1.26,1.45,1.66,2,1,1,0.01,0.0267,0.0367,0.0633,0.08,0.12,0.1533, ...
\ No newline at end of file
+rng_seed,param_id,n_total,mean_work_interactions_child,mean_work_interactions_adult,mean_work_interactions_elderly,daily_fraction_work,work_network_rewire,mean_random_interactions_child,sd_random_interactions_child,mean_random_interactions_adult,sd_random_interactions_adult,mean_random_interactions_elderly,sd_random_interactions_elderly,random_interaction_distribution,child_network_adults,elderly ...
```
Some of the other parameters have changed apart from the one you have added, also `baseline_paramaters_transpose.csv` need to be change in sync.

Please update` baseline_paramaters_transpose.csv `then from the main directory call `./python/transpose_parameters.py `which will populate this file.

shaunkhoo replied:
Completed - updated forked master branch and merged local changes (for manual contact tracing) to master branch to update codebase to latest copy

roberthinch commented on src/interventions.c:
```diff
@@ -534,12 +534,22 @@ void intervention_notify_contacts(
+
+				// We first check if this interaction is manually traceable
+				if( inter->manual_traceable == UNKNOWN )
+					inter->manual_traceable = gsl_ran_bernoulli( rng, params->traceable_interaction_fraction );
```
Shouldn't this be` manual_traceable_interaction_fraction`?

Also, another thing to consider is how best to model this in the context of manual contract tracing. If you see the same person on multiple days, do you have multiple chances to remember them? For the phone we decided yes as it we are modelling the fact someone was not carrying their phone. Also, is the value independent of the type of network? On the random network I suspect the value for manual contact tracing is very low, but on the occupational network it would be much higher.

shaunkhoo replied:
Completed fix - see discussion on latest pull request

roberthinch commented on src/interventions.c:
```diff
@@ -534,12 +534,22 @@ void intervention_notify_contacts(
+				if( inter->app_traceable == UNKNOWN )
+					inter->app_traceable = gsl_ran_bernoulli( rng, params->traceable_interaction_fraction );
+
+				// If it is manually traceable, we trigger the interventions
```
Given the function call is currently identical for manual or app based tracing, can you consolidate in to a single if statement.

shaunkhoo replied:
Completed with slight changes - consolidated into single if statement, removed the extra property in the 'inter' object. Now differentiated by interaction network, and need to retain information about trace type. See other discussion on 'manual_traceable_interaction_fraction'

roberthinch commented on src/interventions.h:
```diff
@@ -48,6 +48,7 @@ void intervention_quarantine_household( model*, individual*, int, int, trace_tok
 void intervention_test_take( model*, individual* );
 void intervention_test_result( model*, individual* );
 void intervention_notify_contacts( model*, individual*, int, trace_token* );
+void intervention_contact_tracing( model*, individual*, int, trace_token* );
```
I can't see this function in `interventions.c,` please can your remove.

shaunkhoo replied:
Completed - sorry for the oversight!

shaunkhoo commented:
Hi Robert, thanks for comments! I've introduced two additional changes on top of the fixes to the issues you highlighted earlier. I also noted that the tests are failing for this change - I'm not familiar with the entire codebase but I believe at least some of it should be because of the substantial changes being made to the tracing and intervention mechanisms. I am happy to help amend the tests if you could point me in the right direction.

**Manual contact tracing efficacy by interaction network**

Following your recommendation, I have incorporated those changes in the most recent update. There are three parameters controlling manual tracing efficacy for each interaction network (random, occupation, household). This is implemented in the initial trace when the traceability of the individual is unknown, and done alongside the digital contact tracing to establish whether the individual is traceable at all.

Regarding the question of whether multiple interactions implies a higher chance of being manually contact traced, this implementation assumes yes. I took this approach because I believe you are more likely to recall repeated interactions, and contact tracers are more likely to uncover corroborating evidence to help you remember. This is ultimately weighted by the interaction network anyway, so you may still be less likely to recall someone you met randomly twice versus someone you met at work once.

Note: I realised that the code assumes that each interaction is only traced once. If the same interaction record is traced again, then the code will skip the checks on whether it was app or manual traced. In turn, the intervention assumes it was a manual trace and thus introduces a delay. The easiest solution seems to be to re-introduce the 'app_traceable' and 'manual_traceable' properties to the interaction record, as I did before, but happy to take any suggestions on how best to do this. Having it on the interaction record will also allow us to include it in the output, which is another benefit I think is worth considering. Let me know and I can make the changes accordingly.

**Delays in manual contact tracing**

I have also added a parameter to delay the interventions that arise as a result of being traced, which (I believe) essentially delays the effect of manual contact tracing. The integer parameter sets the number of days which manual contact tracing requires before a list of interaction records is retrieved. Unfortunately, I am not privy to the data on what the delay is between a positive result and manual contact tracing as I am working on a voluntary basis. However, publicly available data suggests it takes about 4-5 days between the onset of symptoms and isolation, which puts an upper bound on that number (see page 3 [here](https://www.moh.gov.sg/docs/librariesprovider5/local-situation-report/situation-report---7-jun-2020.pdf))

Note: If the same individual was successfully traced through manual and digital contact tracing, I assume digital takes precedence, and no delay occurs in this tracing situation. 

Let me know if these changes work for you, and especially whether I should go ahead to include the 'app_traceable' and 'manual_traceable' properties to the interaction record. Thanks!

roberthinch commented:
Unfortunately the mechanism for delay will not work since `intervention_quarantine_until` only allows immediate quarantining (as coded it simply increases the length of coding).

roberthinch commented:
Equivalent functionality pushed in PR #106
